README
Raspberry Pi Weather Station

Uses temp, light, wind speed, wind direction, humidity, pressure sensors
and sends data to OpenHab using MQTT


This code uses the following libraries and snippets and code:

BMP085 code and libraries:
https://www.john.geek.nz/2012/08/reading-data-from-a-bosch-bmp085-with-a-raspberry-pi/
https://learn.adafruit.com/bmp085/using-the-bmp085-api-v2

MCP3008 C Library:
http://gpcf.eu/projects/embedded/adc/

WiringPi:
http://wiringpi.com/

DHT22 Code:
http://pi.gpcf.eu/projects/embedded/adc/

MQTT:
https://www.eclipse.org/paho/


changelog
v1.2 	22/10/2014	Changed the code code to use Adafruits python script to read the DHT22.
			The original code I used worked but I found  at least every 1-2 weeks it
			would crash with a segfault which I have not been able to fix.
v1.3	30/03/2018	Updated code to send data to OpenHab using MQTT



AdafruitDHT.py
#!/usr/bin/python
# Copyright (c) 2014 Adafruit Industries
# Author: Tony DiCola

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
import sys

import Adafruit_DHT


# Parse command line parameters.
sensor_args = { '11': Adafruit_DHT.DHT11,
				'22': Adafruit_DHT.DHT22,
				'2302': Adafruit_DHT.AM2302 }
if len(sys.argv) == 3 and sys.argv[1] in sensor_args:
	sensor = sensor_args[sys.argv[1]]
	pin = sys.argv[2]
else:
	print 'usage: sudo ./Adafruit_DHT.py [11|22|2302] GPIOpin#'
	print 'example: sudo ./Adafruit_DHT.py 2302 4 - Read from an AM2302 connected to GPIO #4'
	sys.exit(1)

# Try to grab a sensor reading.  Use the read_retry method which will retry up
# to 15 times to get a sensor reading (waiting 2 seconds between each retry).
humidity, temperature = Adafruit_DHT.read_retry(sensor, pin)

# Note that sometimes you won't get a reading and
# the results will be null (because Linux can't
# guarantee the timing of calls to read the sensor).  
# If this happens try again!
if humidity is not None and temperature is not None:
	print '{0:0.1f},{1:0.1f}'.format(temperature, humidity)
	
else:
	print 'Failed to get reading. Try again!'


mcp3008/gpio.h
#include <stdio.h>
/* 
   GPIO.h (C) 2013 Gabriel Perez-Cerezo 
 Permission is hereby granted, free of charge, to any person obtaining
 a copy of this software and associated documentation files (the
 "Software"), to deal in the Software without restriction, including
 without limitation the rights to use, copy, modify, merge, publish,
 distribute, sublicense, and/or sell copies of the Software, and to
 permit persons to whom the Software is furnished to do so, subject to
 the following conditions:
 
 The above copyright notice and this permission notice shall be included
 in all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

*/

const char path[] = "/sys/class/gpio/gpio";
char pinpath[40];


void gpio_init(int pin, char *direction) { // sets the direction of a pin. Allowed values are "in" and "out"
  FILE *f;
  sprintf(pinpath, "%s%d/direction", path,  pin);
  f = fopen(pinpath, "w");
  fprintf(f, "%s", direction);
  fclose(f);
}

void gpio_write(int pin, int value) { // sets the output value of the pin. Allowed values are 1 and 0.
  sprintf(pinpath, "%s%d/value", path, pin);
  FILE *f;
  f = fopen(pinpath, "w");
  fprintf(f, "%d", value);
  fclose(f);
}

int gpio_read(int pin) { // reads input value from the specified GPIO pin. Returns 1 or 0.
  sprintf(pinpath, "%s%d/value", path, pin);
  FILE *f;
  f = fopen(pinpath, "r");
  int x;
  if ((x = fgetc(f))!=EOF){
    x = (x==49); /* Returns 1 if the input value is HIGH. (fgetc returns 49 when the input value is HIGH). Returns 0 if it is low.
		   */
  }else{
    x = -1;        // -1 is returned when there is an error.
  }
  fclose(f);
  return x;       
}
  


lol_dht22/dht22.c
/*
 *      dht22.c:
 *	Simple test program to test the wiringPi functions
 *	Based on the existing dht11.c
 *	Amended by technion@lolware.net
 */

#include "/usr/local/include/wiringPi.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <sys/types.h>
#include <unistd.h>
#include "locking.h"
#include "common.h"

#define MAXTIMINGS 85
#define DHTPIN 7

float h;
float t;
static int dht22_dat[5] = {0,0,0,0,0};


#ifdef DEBUG
	int mydebug(char *debug_info)
	{
        	FILE *fp;
	        fp=fopen("/tmp/debug.log", "a");
        	fprintf(fp, debug_info);
	        fprintf(fp, "\n");
        	fclose(fp);
        	return 0;
	}
#endif


static uint8_t sizecvt(const int read)
{
  /* digitalRead() and friends from wiringpi are defined as returning a value
  < 256. However, they are returned as int() types. This is a safety function */

  if (read > 255 || read < 0)
  {
	#ifdef DEBUG
    		mydebug("Invalid data from wiringPi library\n");
	#endif
    exit(EXIT_FAILURE);
  }
  return (uint8_t)read;
}

static int read_dht22_dat()
{

// keeps crashing somewhere in here

  uint8_t laststate = HIGH;
  uint8_t counter = 0;
  uint8_t j = 0, i;

  dht22_dat[0] = dht22_dat[1] = dht22_dat[2] = dht22_dat[3] = dht22_dat[4] = 0;

 // pull pin down for 18 milliseconds
  pinMode(DHTPIN, OUTPUT);
  digitalWrite(DHTPIN, LOW);
  delay(18);
  // then pull it up for 40 microseconds
  digitalWrite(DHTPIN, HIGH);
  delayMicroseconds(40); 
  // prepare to read the pin
  pinMode(DHTPIN, INPUT);

  // detect change and read data
  for ( i=0; i< MAXTIMINGS; i++) {
    counter = 0;
    while (sizecvt(digitalRead(DHTPIN)) == laststate) {
      counter++;
      delayMicroseconds(1);
      if (counter == 255) {
        break;
      }
    }
    laststate = sizecvt(digitalRead(DHTPIN));

    if (counter == 255) break;

    // ignore first 3 transitions
    if ((i >= 4) && (i%2 == 0)) {
      // shove each bit into the storage bytes
      dht22_dat[j/8] <<= 1;
      if (counter > 16)
        dht22_dat[j/8] |= 1;
      j++;
    }
  }

  // check we read 40 bits (8bit x 5 ) + verify checksum in the last byte
  // print it out if data is good
  if ((j >= 40) && 
      (dht22_dat[4] == ((dht22_dat[0] + dht22_dat[1] + dht22_dat[2] + dht22_dat[3]) & 0xFF)) ) {
//        float t, h;
        h = (float)dht22_dat[0] * 256 + (float)dht22_dat[1];
        h /= 10;
        t = (float)(dht22_dat[2] & 0x7F)* 256 + (float)dht22_dat[3];
        t /= 10.0;
        if ((dht22_dat[2] & 0x80) != 0)  t *= -1;

//    printf("test\n");		
    return 1;
  }
  else
  {

    //printf("Data not good, skip\n");
    return 0;
  }
}

int readDHT22 (void)
{
 int lockfd;

 	#ifdef DEBUG
                mydebug("inside readDHT22()\n");
        #endif


  lockfd = open_lockfile(LOCKFILE);

        #ifdef DEBUG
                mydebug("openlock_file\n");
        #endif



  if (wiringPiSetup () == -1)
	{
	mydebug("wiringpisetup failure/n");
	exit(EXIT_FAILURE) ;
	}

  if (setuid(getuid()) < 0)
  {
	#ifdef DEBUG
		mydebug("Dropping privileges failed\n");
	#endif

    exit(EXIT_FAILURE);
  }

  int i;
  for(i=0; i<10; i++){		// Stop if value not determined in 10 secs
  	while (read_dht22_dat(1) == 0) 
  	{
     		delay(1000); // wait 1sec to refresh
 	}
   }

  close_lockfile(lockfd);

  return 0 ;
  }



BMP085/getBMP085.c
/*
Raspberry Pi Bosch BMP085 communication code.
By:      John Burns (www.john.geek.nz)
Date:    13 February 2013
License: CC BY-SA v3.0 - http://creativecommons.org/licenses/by-sa/3.0/

This is a derivative work based on:
	BMP085 Extended Example Code
	by: Jim Lindblom
	SparkFun Electronics
	date: 1/18/11
	license: CC BY-SA v3.0 - http://creativecommons.org/licenses/by-sa/3.0/
	Source: http://www.sparkfun.com/tutorial/Barometric/BMP085_Example_Code.pde

Compile with: gcc -Wall -o testBMP085 ./smbus.c ./testBMP085.c


Circuit detail:
	Using a Spark Fun Barometric Pressure Sensor - BMP085 breakout board
	link: https://www.sparkfun.com/products/9694
	This comes with pull up resistors already on the i2c lines.
	BMP085 pins below are as marked on the Sparkfun BMP085 Breakout board

	SDA	- 	P1-03 / IC20-SDA
	SCL	- 	P1-05 / IC20_SCL
	XCLR	- 	Not Connected
	EOC	-	Not Connected
	GND	-	P1-06 / GND
	VCC	- 	P1-01 / 3.3V
	
	Note: Make sure you use P1-01 / 3.3V NOT the 5V pin.
*/

#include <stdio.h>
#include <stdint.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <linux/i2c-dev.h>
#include <linux/i2c.h>
#include <sys/ioctl.h>
#include "smbus.h" 

#define BMP085_I2C_ADDRESS 0x77

const unsigned char BMP085_OVERSAMPLING_SETTING = 3;

// Calibration values - These are stored in the BMP085
short int ac1;
short int ac2; 
short int ac3; 
unsigned short int ac4;
unsigned short int ac5;
unsigned short int ac6;
short int b1; 
short int b2;
short int mb;
short int mc;
short int md;

int b5; 

unsigned int temperature, pressure;


// Open a connection to the bmp085
// Returns a file id
int bmp085_i2c_Begin()
{
	int fd;
	char *fileName = "/dev/i2c-0";
	
	// Open port for reading and writing
	if ((fd = open(fileName, O_RDWR)) < 0)
		exit(1);
	
	// Set the port options and set the address of the device
	if (ioctl(fd, I2C_SLAVE, BMP085_I2C_ADDRESS) < 0) {					
		close(fd);
		exit(1);
	}

	return fd;
}

// Read two words from the BMP085 and supply it as a 16 bit integer
__s32 bmp085_i2c_Read_Int(int fd, __u8 address)
{
	__s32 res = i2c_smbus_read_word_data(fd, address);
	if (res < 0) {
		close(fd);
		exit(1);
	}

	// Convert result to 16 bits and swap bytes
	res = ((res<<8) & 0xFF00) | ((res>>8) & 0xFF);

	return res;
}

//Write a byte to the BMP085
void bmp085_i2c_Write_Byte(int fd, __u8 address, __u8 value)
{
	if (i2c_smbus_write_byte_data(fd, address, value) < 0) {
		close(fd);
		exit(1);
	}
}

// Read a block of data BMP085
void bmp085_i2c_Read_Block(int fd, __u8 address, __u8 length, __u8 *values)
{
	if(i2c_smbus_read_i2c_block_data(fd, address,length,values)<0) {
		close(fd);
		exit(1);
	}
}


void bmp085_Calibration()
{
	int fd = bmp085_i2c_Begin();
	ac1 = bmp085_i2c_Read_Int(fd,0xAA);
	ac2 = bmp085_i2c_Read_Int(fd,0xAC);
	ac3 = bmp085_i2c_Read_Int(fd,0xAE);
	ac4 = bmp085_i2c_Read_Int(fd,0xB0);
	ac5 = bmp085_i2c_Read_Int(fd,0xB2);
	ac6 = bmp085_i2c_Read_Int(fd,0xB4);
	b1 = bmp085_i2c_Read_Int(fd,0xB6);
	b2 = bmp085_i2c_Read_Int(fd,0xB8);
	mb = bmp085_i2c_Read_Int(fd,0xBA);
	mc = bmp085_i2c_Read_Int(fd,0xBC);
	md = bmp085_i2c_Read_Int(fd,0xBE);
	close(fd);
}

// Read the uncompensated temperature value
unsigned int bmp085_ReadUT()
{
	unsigned int ut = 0;
	int fd = bmp085_i2c_Begin();

	// Write 0x2E into Register 0xF4
	// This requests a temperature reading
	bmp085_i2c_Write_Byte(fd,0xF4,0x2E);
	
	// Wait at least 4.5ms
	usleep(5000);

	// Read the two byte result from address 0xF6
	ut = bmp085_i2c_Read_Int(fd,0xF6);

	// Close the i2c file
	close (fd);
	
	return ut;
}

// Read the uncompensated pressure value
unsigned int bmp085_ReadUP()
{
	unsigned int up = 0;
	int fd = bmp085_i2c_Begin();

	// Write 0x34+(BMP085_OVERSAMPLING_SETTING<<6) into register 0xF4
	// Request a pressure reading w/ oversampling setting
	bmp085_i2c_Write_Byte(fd,0xF4,0x34 + (BMP085_OVERSAMPLING_SETTING<<6));

	// Wait for conversion, delay time dependent on oversampling setting
	usleep((2 + (3<<BMP085_OVERSAMPLING_SETTING)) * 1000);

	// Read the three byte result from 0xF6
	// 0xF6 = MSB, 0xF7 = LSB and 0xF8 = XLSB
	__u8 values[3];
	bmp085_i2c_Read_Block(fd, 0xF6, 3, values);

	up = (((unsigned int) values[0] << 16) | ((unsigned int) values[1] << 8) | (unsigned int) values[2]) >> (8-BMP085_OVERSAMPLING_SETTING);

	// Close the i2c file
	close (fd);
	
	return up;
}

// Calculate pressure given uncalibrated pressure
// Value returned will be in units of Pa
unsigned int bmp085_GetPressure(unsigned int up)
{
	int x1, x2, x3, b3, b6, p;
	unsigned int b4, b7;
  
	b6 = b5 - 4000;
	// Calculate B3
	x1 = (b2 * (b6 * b6)>>12)>>11;
	x2 = (ac2 * b6)>>11;
	x3 = x1 + x2;
	b3 = (((((int)ac1)*4 + x3)<<BMP085_OVERSAMPLING_SETTING) + 2)>>2;
  
	// Calculate B4
	x1 = (ac3 * b6)>>13;
	x2 = (b1 * ((b6 * b6)>>12))>>16;
	x3 = ((x1 + x2) + 2)>>2;
	b4 = (ac4 * (unsigned int)(x3 + 32768))>>15;
  
	b7 = ((unsigned int)(up - b3) * (50000>>BMP085_OVERSAMPLING_SETTING));
	if (b7 < 0x80000000)
		p = (b7<<1)/b4;
	else
		p = (b7/b4)<<1;
	
	x1 = (p>>8) * (p>>8);
	x1 = (x1 * 3038)>>16;
	x2 = (-7357 * p)>>16;
	p += (x1 + x2 + 3791)>>4;
  
	return p;
}

// Calculate temperature given uncalibrated temperature
// Value returned will be in units of 0.1 deg C
unsigned int bmp085_GetTemperature(unsigned int ut)
{
	int x1, x2;
  
	x1 = (((int)ut - (int)ac6)*(int)ac5) >> 15;
	x2 = ((int)mc << 11)/(x1 + md);
	b5 = x1 + x2;

	unsigned int result = ((b5 + 8)>>4);  

	return result;
}

//int main(int argc, char **argv)
//{
//	bmp085_Calibration();
//	temperature = bmp085_GetTemperature(bmp085_ReadUT());
//	pressure = bmp085_GetPressure(bmp085_ReadUP());
//	
//	printf("Temperature %0.2f ", ((double)temperature)/10);
//	printf("Pressure %0.2f\n", ((double)pressure)/100);
//	return 0;
//}


lol_dht22/locking.c
/*
 *     	locking.c:
 *      Prevents rapid use of application by implementing a file lock
 *      technion@lolware.net
 */

#include <sys/file.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>

#ifdef DEBUG
int ldebug(char *debug_info)
{
        FILE *fp;
        fp=fopen("/tmp/debug.log", "a");
        fprintf(fp, debug_info);
        fprintf(fp, "\n");
        fclose(fp);
        return 0;
}
#endif



int open_lockfile(const char *filename)

{
   int fd;
   fd = open(filename, O_CREAT | O_RDONLY, 0600);

   if (fd < 0)
   {
      	#ifdef DEBUG
        ldebug("Failed to access lock file");
        #endif

	printf("Failed to access lock file: %s\nerror: %s\n",
		filename, strerror(errno));
      exit(EXIT_FAILURE);
   }
   
   while(flock(fd, LOCK_EX | LOCK_NB) == -1)
   {
      if(errno == EWOULDBLOCK)
      {
         printf("Lock file is in use, exiting...\n");
         /* If the lockfile is in use, we COULD sleep and try again.
          * However, a lockfile would more likely indicate an already runaway
	  * process.
         */
	#ifdef DEBUG
        ldebug("Lock file in use");
        #endif

	 exit(EXIT_FAILURE);
      }
      perror("Flock failed");

	#ifdef DEBUG
        ldebug("flock failed");
        #endif

      exit(EXIT_FAILURE); 
   }
   return fd;
}

void close_lockfile(int fd)
{
   if(flock(fd, LOCK_UN) == -1)
   {
      perror("Failed to unlock file");
      exit(EXIT_FAILURE);
   }
   if(close(fd) == -1)
   {
      perror("Closing descriptor on lock file failed");
      exit(EXIT_FAILURE);
   }
}


weather-station.c
/*
Raspberry Pi Weather Station
Sends data to Openhab2 using MQTT

This code uses the following libraries and snippets and code:

BMP085 code and libraries:
https://www.john.geek.nz/2012/08/reading-data-from-a-bosch-bmp085-with-a-raspberry-pi/
https://learn.adafruit.com/bmp085/using-the-bmp085-api-v2

MCP3008 C Library:
http://gpcf.eu/projects/embedded/adc/

WiringPi:
http://wiringpi.com/

*/

#include <wiringPi.h>
#include <wiringPiSPI.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <getopt.h>
#include "BMP085/smbus.c"
#include "BMP085/smbus.h"
#include "BMP085/getBMP085.c"
#include "mcp3008/mcp3008.h"
#include "MQTTClient.h"
#include <time.h>

#define RAIN_PIN 15
#define WIND_PIN 16
#define RAIN_CALIBRATION 0.2794              // 0.2794 mm per tip
#define WIND_CALIBRATION 2.4

#define ADDRESS     "tcp://openhab2.home:1883"
#define CLIENTID    "weatherstation"
#define QOS         1
#define TIMEOUT     10000L
#define TOPIC_temperature       "weather-station/temperature"
#define TOPIC_dewpoint		"weather-station/dewpoint"
#define TOPIC_light		"weather-station/light"
#define TOPIC_uvi          	"weather-station/uvi"
#define TOPIC_rain          	"weather-station/rain"
#define TOPIC_windspeed         "weather-station/windspeed"
#define TOPIC_abs_hum          	"weather-station/abs_hum"
#define TOPIC_pressure		"weather-station/pressure"

float rainCounter;      		// counter for rain guage clicks
float windCounter;
clock_t last_rain_interrupt_time = 0;
clock_t last_wind_interrupt_time = 0;

static const char * optString = "v";
char mystring[50]; 			//size of the number
static const struct option longOpts[] = {
	{ "version", no_argument, NULL, 'v' },
	{ NULL, no_argument,NULL,0}
};


// ======================================================================
// rainInterrupt:  called for rain gauge counter

void rainInterrupt(void) {
        clock_t rain_interrupt_time;
        rain_interrupt_time = clock();

        if (((((float)rain_interrupt_time - (float)last_rain_interrupt_time) / 1000000.0F ) * 1000) >  200)  {
                rainCounter++;
        }
        last_rain_interrupt_time = rain_interrupt_time;
}

// ======================================================================
// windInterrupt

void windInterrupt(void) {
	clock_t wind_interrupt_time;
        wind_interrupt_time = clock();

        if (((((float)wind_interrupt_time - (float)last_wind_interrupt_time) / 1000000.0F ) * 1000) >  5)  {
                windCounter++;
        }
        last_wind_interrupt_time = wind_interrupt_time;

}

//=======================================================================
int read_mcp3008(int channel)
{
	// Channel, Clock, Output, Input, CS 
  int value = mcp3008_value(channel, 11, 9, 10, 8);
  return value;
}

// ======================================================================

#ifdef DEBUG
int debug(char *debug_info)
{
	FILE *fp;
        fp=fopen("/tmp/debug.log", "a");
        fprintf(fp, debug_info);
	fprintf(fp, "\n");
	fclose(fp);
	return 0;
}
#endif

//=======================================================================
float calculate_dew_point(float temp, float rel_humidity)
{
  float a = 17.27;
  float b = 237.7;
  float dp;
  float gamma = ((a * temp)/(b+temp)) + log(rel_humidity/100.0);
  if(a == gamma)
    dp = 0;
  else
    dp= (b*gamma)/(a-gamma);
  return dp;
}

//=======================================================================
float absolute_humidity(float temp, float rh)
{
  float abs_hum = ((6.112 * exp((17.67 * temp)/(temp + 243.5)) * 2.164 * rh) / (273.15 + temp));
  return abs_hum;
}
//=======================================================================
int main(int argc, char **argv)
{

        FILE *ptr_file;			//File pointer for dht22.txt file
        char dht22_buf[20];
	float t;			//DHT22 temp
	float h;			//DHT22 rel humidity
	int result;			//wiringPi result
	float rainFall;
	float windSpeed;


        #ifdef DEBUG
	debug("=================");
	debug("Program Startup");
	debug("=================");
	#endif

	int opt = 0;
        int longIndex = 0;

        opt = getopt_long( argc, argv, optString, longOpts, &longIndex );
        while( opt != -1 ) {
                switch (opt) {
                        case 'v':
                                printf("Version 1.3\n");
                                exit(0);
                        default:
                                exit(0);
                }
                opt = getopt_long( argc, argv, optString, longOpts, &longIndex );
        }

        float start_time, end_time;

        #ifdef DEBUG
		time_t curtime;
                time(&curtime);
                debug(ctime(&curtime));
        #endif

	result = wiringPiSetup ();
        if (result == -1)
        {
                #ifdef DEBUG
			debug("Error on wiringPiSetup.  weatherstation quitting");
		#endif
                return 0;
        }

	//Setup MQTT
	MQTTClient client;
    	MQTTClient_connectOptions conn_opts = MQTTClient_connectOptions_initializer;
    	MQTTClient_message pubmsg = MQTTClient_message_initializer;
    	MQTTClient_deliveryToken token;
    	int rc;
    	MQTTClient_create(&client, ADDRESS, CLIENTID,
        	MQTTCLIENT_PERSISTENCE_NONE, NULL);
    	conn_opts.keepAliveInterval = 20;
    	conn_opts.cleansession = 1;
    	if ((rc = MQTTClient_connect(client, &conn_opts)) != MQTTCLIENT_SUCCESS)
    	{
            printf("Failed to connect, return code %d\n", rc);
            exit(EXIT_FAILURE);
    	}

	#ifdef DEBUG
               debug("Setup MQTT Complete");
        #endif

	//Setup interrupt
	if ( wiringPiISR (RAIN_PIN, INT_EDGE_BOTH, &rainInterrupt) < 0 ) {
                printf("Unable to setup ISR");
        }

	if ( wiringPiISR (WIND_PIN, INT_EDGE_FALLING, &windInterrupt) < 0 ) {
                printf("Unable to setup ISR");
	}


	float dewpoint;
	float abs_humidity;
	float pressure;

	start_time = time(NULL);

	for(; /* some condition that takes forever to meet */;) {
     			// do stuff that apparently takes forever.

		end_time = time(NULL);
       		double diff_time = difftime(end_time, start_time);


    		if (diff_time >= 60) {				// Upload data to internal server 5 mins
                	//printf("5 mins passed - send data...\n");

			start_time = time(NULL);
			#ifdef DEBUG
				time(&curtime);
				debug(ctime(&curtime));
				debug("Reading mcp3008 -chan 0");
			#endif

			int myval = read_mcp3008(0);			//Read UVI
			#ifdef DEBUG
				debug("done chan 0");
			#endif

			float vout = myval/1023.0 * 3.3;
			float sensorVoltage = vout / 471.0;
			float millivolts = sensorVoltage * 1000.0;
			float uvi = millivolts * (5.25/20.0);

			#ifdef DEBUG
				debug("Reading mcp3008 - chan 3");
			#endif

			myval = read_mcp3008(3);				//Read temt6000

			#ifdef DEBUG
				debug("done chan 3");
			#endif

			vout = myval/1023.0 * 3.3;
			float microAmps = vout * 100.0;			// microamps = Vout/10k (resistor on temt6000) * 1000000
			float Light = microAmps * 2;				//Acording to datasheet graph

			#ifdef DEBUG
				debug("Reading dht22");
			#endif

			// readDHT22();				// Read DHT22 Sensor

			//Call Adafruits python script to read dht22 sensor on pin 4
			//Saves result to test file
			system("/usr/bin/python /home/pi/RaspberryPi-WeatherStation/AdafruitDHT.py 22 4 > /tmp/dht22.txt");

			//Open file and get values
			ptr_file =fopen("/tmp/dht22.txt","r");
			while (fgets(dht22_buf,20, ptr_file)!=NULL)
	                       sscanf(dht22_buf, "%f,%f", &t, &h);
			fclose(ptr_file);

			#ifdef DEBUG
				debug("done dht22");
				debug("Reading bmp085");
			#endif
			bmp085_Calibration();

			#ifdef DEBUG
				debug("done bmp085");
        		#endif


			rainFall = rainCounter*RAIN_CALIBRATION;
			windSpeed = windCounter/4*WIND_CALIBRATION;   //4 pulses per rotation

			float temperature = bmp085_GetTemperature(bmp085_ReadUT());
			temperature = temperature / 10.0;

			#ifdef DEBUG
				debug("got temp");
        		#endif

			pressure = bmp085_GetPressure(bmp085_ReadUP());

			#ifdef DEBUG
				debug("got pressure");
			#endif

			dewpoint = calculate_dew_point(t, h);

			#ifdef DEBUG
				debug("calc  dewpoint");
			#endif

			abs_humidity = absolute_humidity(t, h);

			#ifdef DEBUG
				debug("send data to openhab");
			#endif

		        MQTTClient_create(&client, ADDRESS, CLIENTID,
       	 		        MQTTCLIENT_PERSISTENCE_NONE, NULL);
		        if ((rc = MQTTClient_connect(client, &conn_opts)) != MQTTCLIENT_SUCCESS)
        		{
   	        	 printf("Failed to connect, return code %d\n", rc);
 	       		    exit(EXIT_FAILURE);
  		         }

			sprintf(mystring, "%g", temperature);
			pubmsg.payload = mystring; 
    			pubmsg.payloadlen = strlen(mystring);
    			pubmsg.retained = 0;
    			MQTTClient_publishMessage(client, TOPIC_temperature, &pubmsg, &token);
    			rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
			#ifdef DEBUG
	    			debug("Message with delivery:temperature");
				debug(mystring);
			#endif

			sprintf(mystring, "%0.2g", dewpoint);
			pubmsg.payload = mystring; 
                        pubmsg.payloadlen = strlen(mystring);
			MQTTClient_publishMessage(client, TOPIC_dewpoint, &pubmsg, &token);
                        rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
			#ifdef DEBUG
				debug("Message with delivery:dewpoint");
				debug(mystring);
			#endif

			sprintf(mystring, "%g", pressure);
			pubmsg.payload = mystring; 
                        pubmsg.payloadlen = strlen(mystring);
                        MQTTClient_publishMessage(client, TOPIC_pressure, &pubmsg, &token);
                        rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
                        #ifdef DEBUG
				debug("Message with delivery:pressure");
				debug(mystring);
			#endif

 			sprintf(mystring, "%0.2g", Light);
                        pubmsg.payload = mystring; 
                        pubmsg.payloadlen = strlen(mystring);
                        MQTTClient_publishMessage(client, TOPIC_light, &pubmsg, &token);
                        rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
                        #ifdef DEBUG
				debug("Message with delivery:light");
				debug(mystring);
			#endif

			sprintf(mystring, "%0.2g", uvi);
                        pubmsg.payload = mystring; 
                        pubmsg.payloadlen = strlen(mystring);
                        MQTTClient_publishMessage(client, TOPIC_uvi, &pubmsg, &token);
                        rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
                        #ifdef DEBUG
				debug("Message with delivery:uvi");
				debug(mystring);
			#endif

			sprintf(mystring, "%0.3g", abs_humidity);
                        pubmsg.payload = mystring; 
                        pubmsg.payloadlen = strlen(mystring);
                        MQTTClient_publishMessage(client, TOPIC_abs_hum, &pubmsg, &token);
                        rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
                        #ifdef DEBUG
				debug("Message with delivery:abs_hum");
				debug(mystring);
			#endif

                        sprintf(mystring, "%g", windSpeed);
                        pubmsg.payload = mystring; 
                        pubmsg.payloadlen = strlen(mystring);
                        MQTTClient_publishMessage(client, TOPIC_windspeed, &pubmsg, &token);
                        rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
                        #ifdef DEBUG
				debug("Message with delivery:windspeed");
				debug(mystring);
			#endif


                        sprintf(mystring, "%g", rainFall);
                        pubmsg.payload = mystring; 
                        pubmsg.payloadlen = strlen(mystring);
                        MQTTClient_publishMessage(client, TOPIC_rain, &pubmsg, &token);
                        rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
                        #ifdef DEBUG
				debug("Message with delivery:rain");
				debug(mystring);
			#endif


    			MQTTClient_disconnect(client, 10000);
			MQTTClient_destroy(&client);
			}

	}
	return 0;
}

